src/librustc_mir/transform/move_up_propagation.rs

```diff
@@ -0,0 +1,596 @@
+
+pub struct MoveUpPropagation;
+
+impl Pass for MoveUpPropagation {}
+
+impl<'tcx> MirPass<'tcx> for MoveUpPropagation {
+    fn run_pass<'a>(&mut self,
+                    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                    src: MirSource,
+                    mir: &mut Mir<'tcx>) {
+        // we only return when mir_opt_level > 1
+        match tcx.sess.opts.debugging_opts.mir_opt_level {
+            Some(0) |
+            Some(1) |
+            None => { return; },
+            _ => {}
+        };
+        let node_id = src.item_id();
+        let node_path = tcx.item_path_str(tcx.map.local_def_id(node_id));
+        debug!("move-up-propagation on {:?}", node_path);
+
+        let mut opt_counter = 0;
+        let mut dead_temps = BitVector::new(mir.temp_decls.len());
+        let mut dead_vars = BitVector::new(mir.var_decls.len());
+        while let Some((tmp, use_bb, use_idx, def_bb, def_idx)) = get_one_optimization(mir) {
+            let new_statement_kind = get_replacement_statement(mir,
+                                                               use_bb,
+                                                               use_idx,
+                                                               def_bb,
+                                                               def_idx);
+            {
+                let mut bbs = mir.basic_blocks_mut();
+                // replace Def(tmp = ...) with DEST = ...
+                let new_def_stmts: Vec<_> = bbs[def_bb].statements
+                                                        .iter()
+                                                        .enumerate()
+                                                        .map(|(stmt_idx, orig_stmt)| {
+                    if stmt_idx == def_idx {
+                        let new_statement = Statement {
+                            kind: new_statement_kind.clone(),
+                            source_info: orig_stmt.source_info
+                        };
+                        debug!("replacing: {:?} with {:?}.", orig_stmt, new_statement);
+                        new_statement
+                    } else {
+                        orig_stmt.clone()
+                    }
+                }).collect();
+                bbs[def_bb] = BasicBlockData {
+                    statements: new_def_stmts,
+                    terminator: bbs[def_bb].terminator.clone(),
+                    is_cleanup: bbs[def_bb].is_cleanup,
+                };
+
+                // remove DEST = tmp
+                let mut idx_cnt = 0;
+                bbs[use_bb].statements.retain(|orig_stmt| {
+                    let dead = idx_cnt == use_idx;
+                    idx_cnt += 1;
+                    if dead {
+                        debug!("deleting: {:?}", orig_stmt);
+                    }
+                    !dead
+                });
+            }
+
+            // tmp is really a local index (an index into an imaginary vector...)
```

Comment:
Nit: why not `match tmp { ... }` and handle the various cases? seems cleaner than these ifs...